### DISCUSSION TITLE
walkTokens and local context

### DISCUSSION THREAD

**Discussion by j3rem1e** [markedjs/marked#3093] [Q&A] [answered]
Hello,

I have implemented a plugin to transform some elements in the markdown using a "local context": Each markdown is parsed to html using a context linked to the markdown. For example, it transforms links or code blocks using properties of the parent component to contextualize some graphics or href.

In order to do that, I have to create a new Marked instance every time I want to parse the markdown.

My question is mainly about performances:
- Is it costly to create a Marked instance ? I'm worried about the parsing of the configuration of every plugins I need
- If it's costly, how to optimize the creation of instances ? (cloning ?)
- Can I implement a walkTokens using a context without creating a new Marked instance ?

Thanks!

**Comment by UziTech** [comment, marked as answer]
> Is it costly to create a Marked instance?

I don't think we have any metrics on how costly it is to create a marked instance but it shouldn't be too costly.

> If it's costly, how to optimize the creation of instances?

If each instance uses the same extensions you should be able to reuse the instance and just run `marked.parse` again.

> Can I implement a walkTokens using a context without creating a new Marked instance?

Yes. WalkTokens does not have context of parent tokens, but you can create your own version of walk tokens that does.

Something like:

```js
const tokens = marked.lexer(markdown);

// update tokens here

const html = marked.parser(tokens);
```

**Comment by j3rem1e** [comment]
If I call myself lexer/walk/parser, I have to reimplements some logics already handled by marked, like calling hooks, walkTokens of possible extensions, etc, no ?

**Reply by UziTech** [reply]
Yes. This is essentially how `marked.parse` works. https://github.com/markedjs/marked/blob/ba3b7e4be7b02699b44d54a68df74c9f8ac7ad4b/src/Instance.ts#L279-L294
